developed getExposure and setExposure; developed makeGif

AlexKotov opened the pull request:
some new functions that allows to make gif images from multiple images and save it in 480p;
also there is two methods that helps to change "brightness" of canvas and images;
also developed flash effect when taking photo like in native camera app

AlexKotov commented on ios/RCT/RCTCameraManager.m:
```diff
@@ -10,16 +10,20 @@
 
 @implementation RCTCameraManager
 
+@synthesize overlay;
```
i'll remove this and will push to master

AlexKotov commented:
I'll migrate to RNCamera completely then add this features to new lib and
make pull request

2018-04-17 15:54 GMT+03:00 João Guilherme Fidelis <notifications@github.com>
:

> *@jgfidelis* requested changes on this pull request.
>
> Just delete the synthesize.
> Please keep in mind that RCTCamera is deprecated, consider sending PRs to
> RNCamera with these features
> ------------------------------
>
> In ios/RCT/RCTCameraManager.m
> <https://github.com/react-native-community/react-native-camera/pull/1491#discussion_r182058534>
> :
>
> >
>  @end
>
>  @implementation RCTCameraManager
>
> +@synthesize overlay;
>
> no need for this, this was necessary on the early days of ObjC only
> ------------------------------
>
> In ios/RCT/RCTCameraManager.m
> <https://github.com/react-native-community/react-native-camera/pull/1491#discussion_r182058806>
> :
>
> > @@ -698,6 +791,22 @@ - (void)captureStill:(NSInteger)target options:(NSDictionary *)options orientati
>                rotatedCGImage = CGImageCreateWithImageInRect(rotatedCGImage, croppedSize);
>            }
>
> +          // add overlay
> +          self.overlay = [[UIView alloc] initWithFrame:self.view.bounds];
>
> nice
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/react-native-community/react-native-camera/pull/1491#pullrequestreview-112792260>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBIPEEN1EaD45Dq5-LZqRZIZ4oP4XGGks5tpeYBgaJpZM4TXxHs>
> .
>

sibelius commented:
I think you can add to both RCTCamera and RNCamera

could u add this to the docs as well?

AlexKotov commented:
thanx.

2018-04-17 18:08 GMT+03:00 João Guilherme Fidelis <notifications@github.com>
:

> *@jgfidelis* approved this pull request.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/react-native-community/react-native-camera/pull/1491#pullrequestreview-112854266>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBIPKXSMW-h16_HHbyd6WsvUaxjLm6gks5tpgVVgaJpZM4TXxHs>
> .
>

sibelius commented:
@AlexKotov can u add those methods to RNCamera as well?

AlexKotov commented:
I'm doing this now

2018-04-18 16:39 GMT+03:00 Sibelius Seraphini <notifications@github.com>:

> @AlexKotov <https://github.com/AlexKotov> can u add those methods to
> RNCamera as well?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/react-native-community/react-native-camera/pull/1491#issuecomment-382389632>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBIPGWsQVSma6rc_-MBKSSKCqv0VDkHks5tp0IRgaJpZM4TXxHs>
> .
>

AlexKotov commented:
I see that pictures taken by RNCamera "falling" into some separate part of
memory, "file:///var/mobile/Containers/Data/Application/", not in Camera
Roll or Gallery. How can I configure this?

2018-04-18 16:52 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:

> I'm doing this now
>
> 2018-04-18 16:39 GMT+03:00 Sibelius Seraphini <notifications@github.com>:
>
>> @AlexKotov <https://github.com/AlexKotov> can u add those methods to
>> RNCamera as well?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/react-native-community/react-native-camera/pull/1491#issuecomment-382389632>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AFBIPGWsQVSma6rc_-MBKSSKCqv0VDkHks5tp0IRgaJpZM4TXxHs>
>> .
>>
>
>

AlexKotov commented:
Or how can I get access to these pictures from React Native or from your lib

2018-04-18 17:32 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:

> I see that pictures taken by RNCamera "falling" into some separate part of
> memory, "file:///var/mobile/Containers/Data/Application/", not in Camera
> Roll or Gallery. How can I configure this?
>
> 2018-04-18 16:52 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:
>
>> I'm doing this now
>>
>> 2018-04-18 16:39 GMT+03:00 Sibelius Seraphini <notifications@github.com>:
>>
>>> @AlexKotov <https://github.com/AlexKotov> can u add those methods to
>>> RNCamera as well?
>>>
>>> —
>>> You are receiving this because you were mentioned.
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/react-native-community/react-native-camera/pull/1491#issuecomment-382389632>,
>>> or mute the thread
>>> <https://github.com/notifications/unsubscribe-auth/AFBIPGWsQVSma6rc_-MBKSSKCqv0VDkHks5tp0IRgaJpZM4TXxHs>
>>> .
>>>
>>
>>
>

AlexKotov commented:
See this method from official React Native 0.55: saveToCameraRoll()
I can use it for my purposes

2018-04-18 17:35 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:

> Or how can I get access to these pictures from React Native or from your
> lib
>
> 2018-04-18 17:32 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:
>
>> I see that pictures taken by RNCamera "falling" into some separate part
>> of memory, "file:///var/mobile/Containers/Data/Application/", not in
>> Camera Roll or Gallery. How can I configure this?
>>
>> 2018-04-18 16:52 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:
>>
>>> I'm doing this now
>>>
>>> 2018-04-18 16:39 GMT+03:00 Sibelius Seraphini <notifications@github.com>
>>> :
>>>
>>>> @AlexKotov <https://github.com/AlexKotov> can u add those methods to
>>>> RNCamera as well?
>>>>
>>>> —
>>>> You are receiving this because you were mentioned.
>>>> Reply to this email directly, view it on GitHub
>>>> <https://github.com/react-native-community/react-native-camera/pull/1491#issuecomment-382389632>,
>>>> or mute the thread
>>>> <https://github.com/notifications/unsubscribe-auth/AFBIPGWsQVSma6rc_-MBKSSKCqv0VDkHks5tp0IRgaJpZM4TXxHs>
>>>> .
>>>>
>>>
>>>
>>
>

AlexKotov commented:
Used method. Now picture is in Camera Roll but it's rotated to the left
(90) :(

2018-04-18 17:39 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:

> See this method from official React Native 0.55: saveToCameraRoll()
> I can use it for my purposes
>
> 2018-04-18 17:35 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:
>
>> Or how can I get access to these pictures from React Native or from your
>> lib
>>
>> 2018-04-18 17:32 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:
>>
>>> I see that pictures taken by RNCamera "falling" into some separate part
>>> of memory, "file:///var/mobile/Containers/Data/Application/", not in
>>> Camera Roll or Gallery. How can I configure this?
>>>
>>> 2018-04-18 16:52 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:
>>>
>>>> I'm doing this now
>>>>
>>>> 2018-04-18 16:39 GMT+03:00 Sibelius Seraphini <notifications@github.com
>>>> >:
>>>>
>>>>> @AlexKotov <https://github.com/AlexKotov> can u add those methods to
>>>>> RNCamera as well?
>>>>>
>>>>> —
>>>>> You are receiving this because you were mentioned.
>>>>> Reply to this email directly, view it on GitHub
>>>>> <https://github.com/react-native-community/react-native-camera/pull/1491#issuecomment-382389632>,
>>>>> or mute the thread
>>>>> <https://github.com/notifications/unsubscribe-auth/AFBIPGWsQVSma6rc_-MBKSSKCqv0VDkHks5tp0IRgaJpZM4TXxHs>
>>>>> .
>>>>>
>>>>
>>>>
>>>
>>
>

jgfidelis commented:
Have you tried using options like forceUpOrientation?

AlexKotov commented:
forceUpOrientation helped. Thank u.

2018-04-18 19:55 GMT+03:00 João Guilherme Fidelis <notifications@github.com>
:

> Have you tried using options like forceUpOrientation?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/react-native-community/react-native-camera/pull/1491#issuecomment-382456085>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBIPMoWIrAXBi5Syt5HLm0zDeNWGelBks5tp3AagaJpZM4TXxHs>
> .
>

AlexKotov commented:
Think I need help with bringing improvements to RNCamera

I added to RNCameraManager.m lines:

RCT_REMAP_METHOD(getExposureCompensationRange,

                 reactTag:(nonnull NSNumber *)reactTag

                 callback:(RCTResponseSenderBlock)callback) {

  [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager,
NSDictionary<NSNumber *, RNCamera *> *viewRegistry) {

    RNCamera *view = viewRegistry[reactTag];

    if (![view isKindOfClass:[RNCamera class]]) {

      RCTLogError(@"Invalid view returned from registry, expecting
RNCamera, got: %@", view);

    } else {

      callback(@[[NSNull null], [view getExposureCompensationRange]]);

    }

  }];

}

I added to RNCamera.m lines:

- (NSArray *)getExposureCompensationRange

{

  AVCaptureDevice *backCamera = [[self videoCaptureDeviceInput] device];



  float min = backCamera.minExposureTargetBias;

  float max = backCamera.maxExposureTargetBias;



  NSNumber *minNumber = [NSNumber numberWithFloat:min];

  NSNumber *maxNumber = [NSNumber numberWithFloat:max];

  id objects[] = {minNumber, maxNumber};

  NSUInteger count = sizeof(objects) / sizeof(id);



  return [NSArray arrayWithObjects:objects count:count];

}

I added to RNCamera.h line:

- (NSArray *)getExposureCompensationRange;

I changed JS code from:
NativeModules.CameraManager
to:
NativeModules.RNCameraManager

When I'm using getExposureCompensationRange from JS I have an error:
'NativeModules.RNCameraManager.getExposureCompensationRange was called with
1 arguments but expects 2 arguments'

2018-04-19 12:14 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:

> forceUpOrientation helped. Thank u.
>
> 2018-04-18 19:55 GMT+03:00 João Guilherme Fidelis <
> notifications@github.com>:
>
>> Have you tried using options like forceUpOrientation?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/react-native-community/react-native-camera/pull/1491#issuecomment-382456085>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AFBIPMoWIrAXBi5Syt5HLm0zDeNWGelBks5tp3AagaJpZM4TXxHs>
>> .
>>
>
>

jgfidelis commented:
Don't change the JS code, NativeModules.CameraManager, this is already like this for a reason.

How are you calling `getExposureCompensationRange`? Check the methods that are already being called to see how it should be. You should pass the cameraHandle to be the `reactTag` parameter

AlexKotov commented:
If I don't change JS code then wrong methods would be callen (methods from
old lib - RCT, not from new one - RN)

2018-04-20 18:55 GMT+03:00 João Guilherme Fidelis <notifications@github.com>
:

> Don't change the JS code, NativeModules.CameraManager, this is already
> like this for a reason.
>
> How are you calling getExposureCompensationRange? Check the methods that
> are already being called to see how it should be. You should pass the
> cameraHandle to be the reactTag parameter
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/react-native-community/react-native-camera/pull/1491#issuecomment-383140954>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBIPLQFCooPAvqSY9FwE7_7PYT2eh3lks5tqgTvgaJpZM4TXxHs>
> .
>

jgfidelis commented:
what js file are you changing?

AlexKotov commented:
what do u mean?

2018-04-20 19:44 GMT+03:00 João Guilherme Fidelis <notifications@github.com>
:

> what js file are you changing?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/react-native-community/react-native-camera/pull/1491#issuecomment-383154985>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBIPPhRw8inTR3dn7t9FeZ90PfehAXMks5tqhB0gaJpZM4TXxHs>
> .
>

AlexKotov commented:
ok. Now i'm using NativeModules.CameraManager. I had modified RNCamera.js
and index.d.ts but getExposureCompensationRange from RCTCameraManager.m was
called.
I have problem with understanding of bridging in your lib.

2018-04-20 20:51 GMT+03:00 Александр Котов <alexcranston55@gmail.com>:

> what do u mean?
>
> 2018-04-20 19:44 GMT+03:00 João Guilherme Fidelis <
> notifications@github.com>:
>
>> what js file are you changing?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/react-native-community/react-native-camera/pull/1491#issuecomment-383154985>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AFBIPPhRw8inTR3dn7t9FeZ90PfehAXMks5tqhB0gaJpZM4TXxHs>
>> .
>>
>
>

n1ru4l commented:
Going to close this since RCTCamera is deprecated and be removed in the next major release. Please feel free to implement these feature for RNCamera. 

Please also try to put only a single feature in to one pull request to reduce review complexity.
